Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35058 C++ smoketests for the API.
* #34668 Python-driven testing for the new operator registration API
* #34620 Minor OperatorName cleanup
* #34618 Improved error checking
* #34601 Implement schema matching inside Dispatcher
* #34456 Finish implementing schema-less impl, switch all internal registrations to it
* **#34455 Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)**
* #34221 Relax internal invariants of Dispatcher in preparation for schema-less registration

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20330413](https://our.internmc.facebook.com/intern/diff/D20330413)

ezyang committed:
Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20330413](https://our.internmc.facebook.com/intern/diff/D20330413)

[ghstack-poisoned]

ezyang committed:
Update on "Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20330413](https://our.internmc.facebook.com/intern/diff/D20330413)

[ghstack-poisoned]

ezyang committed:
Update on "Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20330413](https://our.internmc.facebook.com/intern/diff/D20330413)

[ghstack-poisoned]

ezyang committed:
Update on "Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20330413](https://our.internmc.facebook.com/intern/diff/D20330413)

[ghstack-poisoned]

ezyang committed:
Update on "Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20330413](https://our.internmc.facebook.com/intern/diff/D20330413)

[ghstack-poisoned]

ezyang committed:
Update on "Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20330413](https://our.internmc.facebook.com/intern/diff/D20330413)

[ghstack-poisoned]

smessmer commented on torch/csrc/jit/frontend/function_schema_parser.cpp:
```diff
@@ -300,5 +300,11 @@ C10_EXPORT FunctionSchema parseSchema(const std::string& schema) {
 
+C10_EXPORT OperatorName parseName(const std::string& name) {
+  auto parsed = parseSchemaOrName(name);
+  TORCH_CHECK(parsed.is_left(), "Tried to parse a name but function schema was given");
```
Tried to parse an **operator** name ?

ezyang committed:
Update on "Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20330413](https://our.internmc.facebook.com/intern/diff/D20330413)

[ghstack-poisoned]

ezyang committed:
Update on "Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20330413](https://our.internmc.facebook.com/intern/diff/D20330413)

[ghstack-poisoned]

ezyang commented:
All CR comments here have been done

ezyang committed:
Update on "Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20330413](https://our.internmc.facebook.com/intern/diff/D20330413)

[ghstack-poisoned]
